File: ios-native/NeobankNative/Sources/Features/Notifications/NotificationsController.swift
import Foundation
import Observation

@MainActor
@Observable
final class NotificationsController {
    enum LoadState {
        case loading
        case loaded(NotificationPage)
        case failed(APIError)
    }

    private(set) var state: LoadState = .loading

    private let repository: NotificationRepository

    init(repository: NotificationRepository) {
        self.repository = repository
    }

    func load() async {
        state = .loading
        do {
            state = .loaded(try await repository.list())
        } catch let error as APIError {
            state = .failed(error)
        } catch {
            state = .failed(.decoding())
        }
    }

    /// Used by the background poll loop: keeps the last good page on screen
    /// instead of flashing a spinner or error every 30 seconds on a blip.
    func refresh() async {
        guard let page = try? await repository.list() else { return }
        state = .loaded(page)
    }

    func markRead(_ id: String) async throws {
        guard case .loaded(var page) = state else { return }
        let updated = try await repository.markRead(id)
        if let index = page.notifications.firstIndex(where: { $0.id == id }) {
            page.notifications[index] = updated
        }
        page.unreadCount = max(0, page.unreadCount - (updated.read ? 1 : 0))
        state = .loaded(page)
    }

    func markAllRead() async throws {
        guard case .loaded(var page) = state else { return }
        try await repository.markAllRead()
        page.notifications = page.notifications.map { $0.withRead(true) }
        page.unreadCount = 0
        state = .loaded(page)
    }
}

#if DEBUG
extension NotificationsController {
    static var preview: NotificationsController {
        NotificationsController(repository: NotificationRepository(client: APIClient(baseURL: AppConfig.apiBaseURL)))
    }
}
#endif

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/notifications/NotificationsController.kt
package dev.horobets.neobanknative.features.notifications

import androidx.compose.runtime.staticCompositionLocalOf
import dev.horobets.neobanknative.core.networking.ApiError
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface NotificationsLoadState {
    data object Loading : NotificationsLoadState
    data class Loaded(val page: NotificationPage) : NotificationsLoadState
    data class Failed(val error: ApiError) : NotificationsLoadState
}

class NotificationsController(private val repository: NotificationRepository) {
    private val _state = MutableStateFlow<NotificationsLoadState>(NotificationsLoadState.Loading)
    val state: StateFlow<NotificationsLoadState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = NotificationsLoadState.Loading
        _state.value = try {
            NotificationsLoadState.Loaded(repository.list())
        } catch (e: ApiError) {
            NotificationsLoadState.Failed(e)
        } catch (e: Exception) {
            NotificationsLoadState.Failed(ApiError.decoding())
        }
    }

    /** Used by the background poll loop: keeps the last good page on screen instead of flashing a spinner or error every 30 seconds on a blip. */
    suspend fun refresh() {
        val page = try {
            repository.list()
        } catch (e: Exception) {
            return
        }
        _state.value = NotificationsLoadState.Loaded(page)
    }

    suspend fun markRead(id: String) {
        val loaded = _state.value as? NotificationsLoadState.Loaded ?: return
        val updated = repository.markRead(id)
        val notifications = loaded.page.notifications.map { if (it.id == id) updated else it }
        val newUnread = (loaded.page.unreadCount - if (updated.read) 1 else 0).coerceAtLeast(0)
        _state.value = NotificationsLoadState.Loaded(loaded.page.copy(notifications = notifications, unreadCount = newUnread))
    }

    suspend fun markAllRead() {
        val loaded = _state.value as? NotificationsLoadState.Loaded ?: return
        repository.markAllRead()
        val notifications = loaded.page.notifications.map { it.copy(read = true) }
        _state.value = NotificationsLoadState.Loaded(loaded.page.copy(notifications = notifications, unreadCount = 0))
    }
}

val LocalNotificationsController = staticCompositionLocalOf<NotificationsController> {
    error("No NotificationsController provided — wrap the composition in AppEnvironment's providers")
}
